src/policy/ancestor_packages.h

```diff
@@ -0,0 +1,167 @@
+                    if (ancestor_subset.size() == rhs.ancestor_subset.size()) {
+                        // Individual feerate. This is not necessarily fee-optimal, but helps in some situations.
+                        // (a.fee * b.vsize > b.fee * a.vsize) is a shortcut for (a.fee / a.vsize > b.fee / b.vsize)
+                        return *fee * *rhs.vsize  > *rhs.fee * *vsize;
+                    }
+                    return ancestor_subset.size() < rhs.ancestor_subset.size();
+                } else {
+                    return mining_sequence.value() < rhs.mining_sequence.value();
+                }
+            }
+        }
+    };
+    /** Map from each txid to PackageEntry */
+    std::map<uint256, PackageEntry> m_txid_to_entry;
+
+    /** Linearized transactions. Always topologically sorted (IsSorted()). If fee information is
+     * provided through LinearizeWithFees(), using mining_sequence scores. */
+    std::vector<std::reference_wrapper<PackageEntry>> m_txns;
+
+    /** Helper function for recursively constructing ancestor caches in ctor. */
+    void visit(const CTransactionRef&);
+public:
+    /** Constructs ancestor package, sorting the transactions topologically and constructing the
+     * txid_to_tx and ancestor_subsets maps. It is ok if the input txns is not sorted.
+     * Expects:
+     * - No duplicate transactions.
+     * - No conflicts between transactions.
+     */
+    AncestorPackage(const Package& txns_in);
+
+    bool IsAncestorPackage() const { return m_ancestor_package_shaped; }
+
+    /** Returns all of the transactions, linearized. */
+    Package Txns() const;
+
+    /** Returns all of the transactions, without the skipped and dangling ones, linearized. */
+    Package FilteredTxns() const;
+
+    /** Get the sorted, filtered ancestor subpackage for a tx. Includes the tx. Does not include
+     * ancestors that are MEMPOOL. If this tx or any ancestors are DANGLING, returns std::nullopt. */
+    std::optional<std::vector<CTransactionRef>> FilteredAncestorSet(const CTransactionRef& tx) const;
+
+    /** Get the total fee and vsize of the ancestor subpackage for a tx. Includes the tx. Does not
+     * include ancestors that are MEMPOOL. If this tx or any ancestors are DANGLING, returns
+     * std::nullopt. This result is always consistent with FilteredAncestorSet(). */
+    std::optional<std::pair<CAmount, int64_t>> FilteredAncestorFeeAndVsize(const CTransactionRef& tx) const;
+
+    /** Get the fee and vsize of a tx. Returns std::nullopt if this information is unknown or this
+     * tx is DANGLING. */
+    std::optional<std::pair<CAmount, int64_t>> GetFeeAndVsize(const CTransactionRef& tx) const;
+
+    /** Should be called if a tx or same txid transaction is found in mempool or submitted to it.
+     * From now on, skip this tx from any result in FilteredAncestorSet(). Does not affect Txns().
+     */
+    void IsInMempool(const CTransactionRef& transaction);
```

Comment:
This sounds like a query, rather a property setter. Perhaps `void SetFoundInMempool(transaction)` to make it clear it's a setter not a getter?

Reply:
Changed to `MarkAsInMempool` to also include the case where it has been found as well as the case where we have submitted it.